lib/src/executable/watch.dart

```diff
@@ -226,34 +182,41 @@ final class _Watcher {
   /// Returns whether all recompilations succeeded.
   Future<bool> _recompileDownstream(Iterable<StylesheetNode> nodes) async {
     var seen = <StylesheetNode>{};
-    var toRecompile = Queue.of(nodes);
-
     var allSucceeded = true;
-    while (toRecompile.isNotEmpty) {
-      var node = toRecompile.removeFirst();
-      if (!seen.add(node)) continue;
-
-      var success = await _compileIfEntrypoint(node.canonicalUrl);
-      allSucceeded = allSucceeded && success;
-      if (!success && _options.stopOnError) return false;
+    while (nodes.isNotEmpty) {
+      nodes = nodes.where((node) => seen.add(node)).toList();
+
+      var sourcesToDestinationsPairs =
+          _sourceEntrypointsToDestinationsPairs(nodes);
+      if (sourcesToDestinationsPairs.isNotEmpty) {
+        var success = await compileStylesheets(
+            _options, _graph, sourcesToDestinationsPairs,
+            ifModified: true);
+        if (!success && _options.stopOnError) return false;
+
+        allSucceeded = allSucceeded && success;
+      }
 
-      toRecompile.addAll(node.downstream);
+      nodes = nodes.expand((node) => node.downstream).toList();
     }
     return allSucceeded;
   }
 
-  /// Compiles the stylesheet at [url] to CSS if it's an entrypoint that's being
-  /// watched.
-  ///
-  /// Returns `false` if compilation failed, `true` otherwise.
-  Future<bool> _compileIfEntrypoint(Uri url) async {
-    if (url.scheme != 'file') return true;
-
-    var source = p.fromUri(url);
-    return switch (_destinationFor(source)) {
-      var destination? => await compile(source, destination),
-      _ => true
-    };
+  /// Returns a sourcesToDestinations mapping for nodes that are entrypoints.
+  Iterable<(String, String)> _sourceEntrypointsToDestinationsPairs(
+      Iterable<StylesheetNode> nodes) {
+    var pairs = <(String, String)>[];
+    for (var node in nodes) {
+      var url = node.canonicalUrl;
+      if (url.scheme != 'file') continue;
+
+      var source = p.fromUri(url);
+      var destination = _destinationFor(source);
+      if (destination != null) {
```

Comment:
In Dart 3, you can check for null and assign in a single statement:

```suggestion
      if (_destinationFor(source) case var destination?) {
```